### PR TITLE
fix(cli): Control clap colors through config

### LINKED ISSUE #9012: Cargo doesn't respect `--color` ~~or `CARGO_TERM_COLOR`~~ for help text and errors from clap
**Problem**

^title

I expect that with `--color=always` option (or `CARGO_TERM_COLOR=always` environment variable) the output will contain expected ANSI escape codes.

**Steps**
1. Run `cargo build --color=always --unknown &> out.txt`
2. The`out.txt` will contain the following output (without ANSI escape codes):

```
error: Found argument '--unknown' which wasn't expected, or isn't valid in this context

USAGE:
    cargo build --color <WHEN>

For more information try --help

```

**Notes**

Output of `cargo version`:

```
cargo 1.50.0-nightly (a3c2627fb 2020-12-14)
```

### PR DESCRIPTION
**Pull request by epage**
### What does this PR try to resolve?

Part of #9012

### How should we test and review this PR?

To accomplish this, I pivoted in how we handle `-C`.  In #11029, I made the config lazily loaded.  Instead, we are now reloading the config for `-C` like we do for "cargo script" and `cargo install`.  If there is any regression, it will be felt by those commands as well and we can fix all together.  As this is unstable, if there is a regression, the impact is less.  This allowed us access to the full config for getting the color setting, rather than taking more limited approaches like checking only `CARGO_TERM_CONFIG`.

### CODE REVIEW

File: src/bin/cargo/cli.rs [hunk 1]
```diff
@@ -40,6 +38,7 @@ pub fn main(lazy_gctx: &mut LazyContext) -> CliResult {
             .into());
         }
         std::env::set_current_dir(&new_cwd).context("could not change to requested directory")?;
+        lazy_gctx.get_mut().reload_cwd()?;
```
[the remark this answers is not in the document]
Comment (weihanglo) [reply src/bin/cargo/cli.rs]: This is not a blocker btw, given `-C` is unstable.

**Review by weihanglo — COMMENTED** [review]
With this patch, the color control works for `CARGO_TERM_COLOR` environment variable and `config.toml`. However, it doesn't work with either `--color never` or `--config 'term.color="never"'`, so I don't think #9012 is resolved.

Not sure if we should ship a partial solution or hold on until `--color` and `--config` get addressed.

File: src/bin/cargo/cli.rs [hunk 2]
```diff
@@ -514,7 +513,19 @@ impl GlobalArgs {
     }
 }
 
-pub fn cli() -> Command {
+pub fn cli(gctx: &GlobalContext) -> Command {
+    // Don't let config errors get in the way of parsing arguments
+    let term = gctx.get::<TermConfig>("term").unwrap_or_default();
```
Comment (weihanglo) [inline src/bin/cargo/cli.rs:518]: The issue is that we sidestep without calling `GlobalContext::merge_cli_args`, which `GlobalContext::configure` does. 

There might also be a discrepancy when we call `reload_cwd()`, but that is minimal as the code path eventually calls `GlobalContext::configure`.

**Comment by epage** [comment]
> Not sure if we should ship a partial solution or hold on until --color and --config get addressed.

I don't see a reasonable way for addressing that part.  We need to be able to parse the command line to get those which requires knowing what level of styling to use.  We could say "parse the command line, take the result, and strip the message based on our own decision".  However, `--help` and errors short-circuits CLI parsing.  We'd have to parse, see if an error gets back, and then re-parse using the "ignore errors" setting which is a very brittle feature in clap.

File: src/bin/cargo/cli.rs (hunk 2 above)
Comment (epage) [reply src/bin/cargo/cli.rs:518]: I'm not too sure what the concern is here.

Is it that we aren't respecting `--color` which you raised elsewhere and you are just calling out the relevant code?

Or is it that we aren't respecting the config from `-C` to control the colors from CLI parsing?  If so, that is a similar problem to `--color` support (and more difficult / brittle to support)

Comment (weihanglo) [reply src/bin/cargo/cli.rs:518]: I was just making sure we're aware of this and also #9012 is not completely resolved. Should we reword #9012 to track `--config` and `--color` or open a new one?

I am good with merge this PR as-is btw. Feel free to `r=weihanglo`.

**Review by weihanglo — APPROVED** [review]

**Comment by weihanglo** [comment]
Seems like we can leverage <https://github.com/rust-lang/cargo/pull/13461> to test colors for this PR, right?

**Comment by weihanglo** [comment]
Okay I didn't notice you've linked the original issue in that PR.

**Pull request merged**